### PR TITLE
EnumType support in flytekit

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This PR introduces Enum Types in flytekit (flyteidl will natively support it)

Example usage:

```python
from enum import Enum

class Color(Enum):
   RED = "red"
   BLUE = "blue"
   GREEN = "green"

@task
def foo(c: Color) -> str:
   return c.value
```

UI/UX: Drop down. RED will be the default value
flytectl: enforce the value to be limited to one of the enum values

Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
[Design Doc](https://docs.google.com/document/d/1rf0hMKbPebJ7jGYpsTDAbJ4ka58KIKD7EgWAEBIH06Q/edit#)

## Tracking Issue
https://github.com/lyft/flyte/issues/590

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -300,12 +301,39 @@ def register(cls, transformer: TypeTransformer):
...
+            find a transformer that matches the generic type of v. e.g List[int], Dict[str, int] etc
+
+        Step 3:
+            if v is of type data class, use the dataclass transformer
+
+        Step 4:
+            Walk the inheritance hierarchy of v and  find a transformer that matches the first base class.
+            This is potentially non-deterministic - will depend on the registration pattern.
+
+            TODO lets make this deterministic by using an ordered dict
```
Comment (cosmicBboy) [inline flytekit/core/type_engine.py:323]: aren't python 3.6+ dicts ordered?

**Review by cosmicBboy — APPROVED** [review]

File: flytekit/core/type_engine.py (hunk 1 above)
Comment (kumare3) [reply flytekit/core/type_engine.py:323]: ohh really? let me check on that

**Review by wild-endeavor — APPROVED** [review]

**Review by wild-endeavor — APPROVED** [review]
